File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/AllergensWrapper.java
package openfoodfacts.github.scrachx.openfood.models;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;

import java.util.ArrayList;
import java.util.List;

import openfoodfacts.github.scrachx.openfood.network.deserializers.AllergensWrapperDeserializer;

/**
 * JSON from URL https://ssl-api.openfoodfacts.org/data/taxonomies/allergens.json (top 14 allergens and substances)
 *
 * @author Lobster
 * @author ross-holloway94 2018-03-14
 */

@JsonDeserialize(using = AllergensWrapperDeserializer.class)
public class AllergensWrapper {

    private List<AllergenResponse> allergens;

    /**
     * @return A list of Allergen objects
     */
    public List<Allergen> map() {
        List<Allergen> entityAllergens = new ArrayList<>();
        for (AllergenResponse allergen : allergens) {
            entityAllergens.add(allergen.map());
        }

        return entityAllergens;
    }

    public void setAllergens(List<AllergenResponse> allergens) {
        this.allergens = allergens;
    }

    public List<AllergenResponse> getAllergens() {
        return allergens;
    }
}

File: app/src/main/java/openfoodfacts/github/scrachx/openfood/models/entities/allergen/AllergensWrapper.kt
package openfoodfacts.github.scrachx.openfood.models.entities.allergen

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import openfoodfacts.github.scrachx.openfood.models.entities.EntityResponse
import openfoodfacts.github.scrachx.openfood.models.entities.EntityWrapper

/**
 * JSON from URL https://ssl-api.openfoodfacts.org/data/taxonomies/allergens.json (top 14 allergens and substances)
 *
 * @author Lobster
 * @author ross-holloway94 2018-03-14
 */
@JsonDeserialize(using = AllergensWrapperDeserializer::class)
class AllergensWrapper(allergens: List<EntityResponse<Allergen>>) : EntityWrapper<Allergen>(allergens)